### PR TITLE
add support for dependencies

### LINKED ISSUE #58: Add dependencies between components
When importing the generated SBOM in dependency-track, there is no dependency tree available. This seems to be because there is no dependency information available in the SBOM. Could this be added to the SBOM?

### PR DESCRIPTION
**Pull request by fnxpt**
Fixes #58 

Still need to add tests for this

### CODE REVIEW

File: lib/cyclonedx/cocoapods/bom_builder.rb [hunk 1]
```diff
@@ -136,6 +137,16 @@ def bom(version: 1)
                 pod.add_to_bom(xml)
               end
             end
+
+            xml.dependencies do
+              @dependencies.each do |key, array|
```
Comment (macblazer) [inline lib/cyclonedx/cocoapods/bom_builder.rb:142]: If `dependencies` is nil, this crashes the app.  Easily seen by running the unit tests with `rake spec`.

File: lib/cyclonedx/cocoapods/bom_builder.rb [hunk 2]
```diff
@@ -119,10 +120,10 @@ def add_to_bom(xml)
     class BOMBuilder
       NAMESPACE = 'http://cyclonedx.org/schema/bom/1.4'
 
-      attr_reader :component, :pods
+      attr_reader :component, :pods, :dependencies
 
-      def initialize(component: nil, pods:)
-        @component, @pods = component, pods
+      def initialize(component: nil, dependencies: nil, pods:)
```
Comment (macblazer) [inline lib/cyclonedx/cocoapods/bom_builder.rb:125]: Codacy is asking you to put `pods:` as the first parameter because `pods` is the only parameter that does not specify a default value.  Calls to this function can then look like one of the following (note that `pods` is always first because it has to be specified):

```
BOMBuilder.new(pods: pods)
BOMBuilder.new(pods: pods, component: myComponent)
BOMBuilder.new(pods: pods, dependencies: deps)
BOMBuilder.new(pods: pods, component: myComponent, dependencies: deps)
```

**Review by macblazer — CHANGES_REQUESTED** [review]
Thanks for working on this.  It does need some unit tests to ensure 

1. The analyzer is finding the right dependencies, and
2. The `BOMBuilder` is generating the right output with the dependencies.

The existing SimplePod and TestingPod fixtures both have a single dependency in them that can help with the tests.  It might be good to create one more fixture with maybe a couple of dependencies to ensure it's iterating over thing correctly.

**Comment by fnxpt** [comment]
Sorry for the delay, I added tests for pod analyzer I will do the same with bom builder later, but for the restrictedPod the code is breaking on the sources, any idea why?

**Comment by macblazer** [comment]
> Sorry for the delay, I added tests for pod analyzer I will do the same with bom builder later, but for the restrictedPod the code is breaking on the sources, any idea why?

The RestrictedPod is a fixture because it has a nice edge case where the Podfile is for iOS, and the `EFQRCode` pod has a dependency that is not used for iOS.  The way CocoaPods works in this case is that the Podfile.lock shows the dependency on `swift_qrcodejs` under the PODS section, but `swift_qrcodejs` is not listed as a top-level pod in the PODS section and is not present at all in the DEPENDENCIES, SPEC REPOS, or SPEC CHECKSUMS sections.  The function `append_all_pod_dependencies` in the podfile_analyzer.rb takes this into account, but your new function `dependencies_hash_of_lockfile_pods` does not.

In the bom.xml output tests for the RestrictedPod, I would not expect to see any references to `swift_qrcodejs` because it is not actually being used for this iOS target.

It might be better to restructure the new code by adding new capability to the `append_all_pod_dependencies` function (at line 169) instead of having the whole new function that iterates over the entire lock file a second time.  Something like

```
def append_all_pod_dependencies(pods_used, pods_cache)
  result = pods_used
  dependencies_hash = [] # new
  original_number = 0
  # Loop adding pod dependencies until we are not adding any more dependencies to the result
  # This brings in all the transitive dependencies of every top level pod.
  # Note this also handles two edge cases:
  #  1. Having a Podfile with no pods used.
  #  2. Having a pod that has a platform-specific dependency that is unused for this Podfile.
  while result.length != original_number
    original_number = result.length
    pods_used.each { |pod_name|
      if pods_cache.key?(pod_name) && !pods_cache[pod_name].empty? # new
         result.push(*pods_cache[pod_name]) # changed
         dependencies_hash[pod_name] = *pods_cache[pod_name] # new
         # maybe additional dependency processing needed here???
      end
    }
    result = result.uniq
    # maybe additional dependency processing needed here???
    pods_used = result
  end
  result # change this to return result and dependencies_hash or make dependencies_hash a class property
end

**Comment by fnxpt** [comment]
I will change it to reflect that

**Comment by fnxpt** [comment]
Fixed the issue with Reserved Podfile... @macblazer I think the only thing missing is adding tests to the bom builder. Are you able to help with this part?

**Comment by fnxpt** [comment]
cleaned up a little bit the code, following your advice, also added bom_builder tests following the already existing tests

**Comment by fnxpt** [comment]
Everything should be ok now

File: spec/cyclonedx/cocoapods/bom_builder_spec.rb [hunk 3]
```diff
@@ -348,13 +363,35 @@
...
         end
+
+        it 'should generate a child dependencies node' do
+          expect(xml.at('bom/dependencies')).not_to be_nil
+        end
+
+        it 'shoudl properly set dependencies node' do
+          dependencies_generated_from_bom_builder = xml.at('bom/dependencies')
+
+          dependencies = Nokogiri::XML(Nokogiri::XML::Builder.new(encoding: 'UTF-8') do |xml|
```
Comment (macblazer) [inline spec/cyclonedx/cocoapods/bom_builder_spec.rb:381]: This look like it's just copied the dependency builder out of bom_builder.rb and placed it here.  This doesn't test the code, but simply tests that you copy/pasted it correctly.

A better test would be something that has hardcoded dependencies as a string here and compare the generated xml to that.  Something like comparing to this string:

```ruby
'<dependencies><dependency ref="pkg:cocoapods/Alamofire@5.6.2"/><dependency ref="pkg:cocoapods/MSAL@1.2.1"><dependency ref="pkg:cocoapods/MSAL@1.2.1#app-lib"/></dependency><dependency ref="pkg:cocoapods/FirebaseAnalytics@7.10.0#app-lib"/><dependency ref="pkg:cocoapods/RxSwift@5.1.2#app-lib"/><dependency ref="pkg:cocoapods/Realm@5.5.1#app-lib"/></dependencies>'
```

Comment (fnxpt) [reply spec/cyclonedx/cocoapods/bom_builder_spec.rb:381]: I tough the same let me replace it

Comment (fnxpt) [reply spec/cyclonedx/cocoapods/bom_builder_spec.rb:381]: well you cannot use it like this because this tests will run twice, one for component without pods and the other with pods... so the first one is just 
`<dependencies/>`

Comment (fnxpt) [reply spec/cyclonedx/cocoapods/bom_builder_spec.rb]: Added the xml strings directly on the tests.

**Comment by fnxpt** [comment]
@macblazer cab you have another look?

File: spec/cyclonedx/cocoapods/bom_builder_spec.rb [hunk 4]
```diff
@@ -348,25 +363,49 @@
...
       let(:component) { CycloneDX::CocoaPods::Component.new(name: 'Application', version: '1.3.5', type: 'application') }
-      let(:bom_builder) { described_class.new(component: component, pods: pods) }
+      let(:bom_builder) { described_class.new(component: component, pods: pods, dependencies: dependencies) }
+      let(:dependencies_result) do
+        '<dependencies>
+                                    <dependency ref="pkg:cocoapods/Alamofire@5.6.2"/>
+                                    <dependency ref="pkg:cocoapods/MSAL@1.2.1">
+                                      <dependency ref="pkg:cocoapods/MSAL@1.2.1#app-lib"/>
+                                    </dependency>
+                                    <dependency ref="pkg:cocoapods/FirebaseAnalytics@7.10.0#app-lib"/>
```
Comment (macblazer) [inline spec/cyclonedx/cocoapods/bom_builder_spec.rb:404]: Should these three purl expressions (lines 404-406) have `#app-lib` at the end of them?  It looks like that is a proper dependency purl for `MSAL/app-lib` but not the others.  The same question holds for new lines 272-274 above.  Looks like maybe something is being used in a loop, and being set once by the `MSAL/app-lib` pod and then being used incorrectly on the next three dependencies.

I don't see right now where that's happening, but I will try to take some time and determine why these show up.  Could be a problem in existing code, or might be a problem with the new dependency analysis code.

Comment (fnxpt) [reply spec/cyclonedx/cocoapods/bom_builder_spec.rb]: @macblazer  fixed the problem, the issue was the dependencies declaration on the spec.

**Review by macblazer — APPROVED** [review]
Thanks for this contribution and working through all the updates.
